sycl/test/sub_group/attributes.cpp

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "helper.hpp"
+#include <CL/sycl.hpp>
+
+#define KERNEL_FUNCTOR_WITH_SIZE(SIZE)                                         \
+  class KernelFunctor##SIZE {                                                  \
+  public:                                                                      \
+    [[cl::intel_reqd_sub_group_size(SIZE)]] void                               \
+    operator()(cl::sycl::nd_item<1> Item) {                                    \
+      const auto GID = Item.get_global_id();                                   \
+    }                                                                          \
+  };
+
+KERNEL_FUNCTOR_WITH_SIZE(1);
+KERNEL_FUNCTOR_WITH_SIZE(2);
+KERNEL_FUNCTOR_WITH_SIZE(4);
+KERNEL_FUNCTOR_WITH_SIZE(8);
+KERNEL_FUNCTOR_WITH_SIZE(16);
+
+#undef KERNEL_FUNCTOR_WITH_SIZE
+
+inline uint32_t flp2(uint32_t X) {
+  X = X | (X >> 1);
+  X = X | (X >> 2);
+  X = X | (X >> 4);
+  X = X | (X >> 8);
+  X = X | (X >> 16);
+  return X - (X >> 1);
+}
+
+int main() {
+  queue Queue;
+  device Device = Queue.get_device();
+
+  // According to specification, this kernel query requires `cl_khr_subgroups`
+  // or `cl_intel_subgroups`, and also `cl_intel_required_subgroup_size`
+  if ((!Device.has_extension("cl_intel_subgroups") &&
+       !Device.has_extension("cl_khr_subgroups")) ||
+      !Device.has_extension("cl_intel_required_subgroup_size")) {
+    std::cout << "Skipping test\n";
+    return 0;
+  }
+
+  try {
+    const auto SGSizes = Device.get_info<info::device::sub_group_sizes>();
+
+    for (const auto SGSize : SGSizes) {
+      // Get the previous power of 2
+      auto ReqdSize = flp2(SGSize);
+
+      cl::sycl::program Prog(Queue.get_context());
+
+      // Store the `cl::sycl::kernel` into a vector because `cl::sycl::kernel`
+      // doesn't have default constructor
+      cl::sycl::vector_class<cl::sycl::kernel> TheKernel;
+
+      switch (ReqdSize) {
+      case 16:
+        Prog.build_with_kernel_type<KernelFunctor16>();
+        TheKernel.push_back(Prog.get_kernel<KernelFunctor16>());
+        break;
+      case 8:
+        Prog.build_with_kernel_type<KernelFunctor8>();
+        TheKernel.push_back(Prog.get_kernel<KernelFunctor8>());
+        break;
+      case 4:
+        Prog.build_with_kernel_type<KernelFunctor4>();
+        TheKernel.push_back(Prog.get_kernel<KernelFunctor4>());
+        break;
+      case 2:
+        Prog.build_with_kernel_type<KernelFunctor2>();
+        TheKernel.push_back(Prog.get_kernel<KernelFunctor2>());
+        break;
+      case 1:
+        Prog.build_with_kernel_type<KernelFunctor1>();
+        TheKernel.push_back(Prog.get_kernel<KernelFunctor1>());
+        break;
+      default:
+        throw feature_not_supported("sub-group size is not supported");
+      }
+
+      auto Kernel = TheKernel[0];
```

Comment:
I see... You are emulating a C++17 `std::optional` since you are in C++11... :-(
On the other hand you could make a specification proposal to add a default constructor to a `cl::sycl::kernel` and a member function to test if the kernel is valid... :-)

Reply:
Not sure whether it makes sense to request it. :)